dali/operators/math/normalize/normalize.cc

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/math/normalize/normalize.h"
+#include "dali/kernels/reduce/reduce.h"
+#include "dali/kernels/normalize/normalize_cpu.h"
+#include "dali/core/math_util.h"
+#include "dali/operators/math/normalize/normalize_utils.h"
+
+namespace dali {
+
+DALI_SCHEMA(Normalize)
+  .DocStr(R"(Normalizes the input by removing mean and dividing by standard deviation.
+
+The mean and standard deviation can be calculated internally for specified subset of axes or
+can be externally provided as *mean* and *stddev* arguments.)")
+  .NumInput(1)
+  .NumOutput(1)
+  .AddOptionalArg("batch", "If True, the mean and standard deviation are calculated across tensors "
+    "in the batch. This also requires that the input sample shapes in the non-averaged axes match.",
+    false)
+  .AddOptionalArg<float>("mean", "Mean value to subtract from the data. It can be either a scalar "
+    "or a batch of tensors with same dimensionality as the input and the shape in each dimension "
+    "must either match that of the input or be equal to 1 (in which case the value will be "
+    "broadcast in this dimension). If not specified, the mean is calculated from the input.",
+    0.0f, true)
+  .AddOptionalArg<float>("stddev", "Stanrad deviation value to scale the data. For shape "
```

Comment:
It is a bit confusing that `stddev` is called scale in the code, while `scale` is a different thing here.

Reply:
NormalizeCPU kerenel can be used for different kind of normalization (base/range, for example), not just mean/stddev - and even if using this kind of normalization, the kernel expects standard deviation's reciprocal. Therefore, the kernel has a more general scale. This operator fuses the scaling factor, so that for each data position X and corresponding parameter position Y the resulting value is:
`out[X] = (in[X] - mean[Y]) * (scale * (stddev[Y] ? 1/stddev[Y] : 0)) + shift`